### DISCUSSION TITLE
Config for command used to list directory content

### DISCUSSION THREAD

**Discussion by kidonng** [PatrickF1/fzf.fish#101] [Archive] [answered]
This plugin currently [uses `ls -A` for listing directory content](https://github.com/PatrickF1/fzf.fish/blob/9a2ead5742f34bfb073f44ccc700c8600a317a22/functions/__fzf_preview_file.fish#L7), could you please add a config like `FZF_PREVIEW_DIR_CMD` in [jethrokuan/fzf](https://github.com/jethrokuan/fzf)? This allows using a different program such as [`exa`](https://github.com/ogham/exa), [`lsd`](https://github.com/Peltoche/lsd), [`colorls`](https://github.com/athityakumar/colorls) and such to list directory content.

This is what it may look like with `exa -a --color=always`:

![image](https://user-images.githubusercontent.com/44045911/102070065-93e13500-3e39-11eb-8394-c1654d5d6369.png)

And with `ls -A`:

![image](https://user-images.githubusercontent.com/44045911/102070139-ac514f80-3e39-11eb-9593-da14059ddfd6.png)

**Comment by PatrickF1** [comment]
Hi @kidonng, thanks for the suggestion. I've been against configurability/extensibility of fzf.fish since I envisioned it. Instead, I favor simple but opinionated code, installation, documentation, plug-and-play functionality, which ultimately means less work for everyone and a lot less work for me as the only maintainers of the plugin. 

However, I am starting to change my mind. Both the `ls` and `fd` invocations can be configured without too much more work. I'm actually thinking about overhauling configurability for v5, but that's not to come for a few weeks at least. I'll keep this open to remind me.

tl;dr sorry not right now but it's not a bad idea and I'm probably going to do it!

**Comment by stephan-hesselmann-by** [comment]
> Hi @kidonng, thanks for the suggestion. I've been against configurability/extensibility of fzf.fish since I envisioned it. Instead, I favor simple but opinionated code, installation, documentation, plug-and-play functionality, which ultimately means less work for everyone and a lot less work for me as the only maintainers of the plugin.

Note that opinionated code can cause breakage and shifts the work to users, because they need to build around the limitations of the plugin. This makes the plugin more brittle and inflexible. For example, the PR https://github.com/PatrickF1/fzf.fish/pull/67 breaks preview for users who link `ls` to `exa`.

**Comment by PatrickF1** [comment]
@stephan-hesselmann-by true. I've always admired opinionated software but perhaps this is the wrong type of software to be opinionated.
Btw, did #67 breaak previews for you? A shortterm fix is to use `command ls`

**Comment by stephan-hesselmann-by** [comment]
I think shell plugins in general will attract users who like to tinker a bit. So while I like sensible defaults to get "plug-and-play functionality", I also appreciate some flexibility in configuration. So I like the idea of adding something like `FZF_PREVIEW_DIR_CMD` with supported default values.

**Comment by PatrickF1** [comment]
Very true: I can expect fish plugin users to want to tinker. Good job, Stephan, you've convinced me to change my vision for this plugin in the span of 10 minutes.
Alright, that's what I'll be working on for v5: greater configurability. Please give me some time to work on it though. I won't release just configuring `ls` or `fd` by themselves because I need to think about it in the context of making everything.
And if there are more things you want configurable, let me know please!

**Comment by kidonng** [comment]
> For example, the PR https://github.com/PatrickF1/fzf.fish/pull/67 breaks preview for users who link `ls` to `exa`.

Pardon my rudeness, but this is a bad example. exa's website clearly states that it is not a drop-in replacement for ls. And modifying coreutils is generally not a good idea. 

That said, I totally agree with your opinion. Having a strong dependence on "third-party" programs (except fzf itself)  may lead to more breakages.

**Comment by dessalines** [comment]
another thing that would be nice to have is not just `FZF_DEFAULT_OPS`, but `FZF_DEFAULT_COMMAND` too

So we could do things like 

`export FZF_DEFAULT_COMMAND='fd --type f'`

Related to #66 as I also ran into that issue.

**Comment by PatrickF1** [comment]
Hey @dessalines, can you clarify what you want to do with FZF_DEFAULT_COMMAND?

**Comment by dessalines** [comment]
Mainly turn off `--follow`

**Comment by PatrickF1** [comment]
Hmm what's tough about using `FZF_DEFAULT_COMMAND` is that I can't assume the user is using it to find files. They could easily hook it up to browse through a directory of cat pictures and then that would mess up the find file functionality.
However, I think I see now that I shouldn't be using `--follow` so I'll turn that off.
EDIT: see ac7308da70a7ab750ef0136974abccc233833f79

**Comment by PatrickF1** [comment, marked as answer]
Hi @kidonng @stephan-hesselmann-by @dessalines, I'm sorry to backtrack but I've changed my mind about making search files more configurable. 

What happened was I discovered CLI file managers and I realized they solve a very similar problem of navigating directories, previewing files, and opening them a million times over and in ways better than fzf.fish ever can without me making it a full time job. There are just so many options available I don't really want to add another one into the mix and compete with them when I can just outsource work to them:
- ranger https://github.com/ranger/ranger
- nnn https://github.com/jarun/nnn
- fff https://github.com/dylanaraps/fff
- lf https://github.com/gokcehan/lf
- vimf https://github.com/vifm/vifm
- broot https://dystroy.org/broot/
- midnight commander https://github.com/MidnightCommander/mc
- noice (https://git.2f30.org/noice/files.html)
- and many, many more viable options, esp. if you use Emacs or Vim
As for myself, I will be using `nnn` more and more and my search files feature less and less.

I know @kidonng said they can't integrate as well as a native fish plugin but I think the benefits they provide tend to outweigh the benefits that a native fish plugin can.

Another reason I changed my mind is I want to be able to focus on adding features other than configurability into fzf.fish. Configurability might be easy to add, but it can double time for testing as I'll need to test both the default path and the configured path. It can also lead to a lot of questions/bug reports from users caused by misconfiguration which will also sap at my time.

Anyway, however, I realized adding support for exa isn't a big deal and makes sense since exa is so popular, nearly as popular as fd, so I will probably do that.

**Reply by PatrickF1** [reply]
Hi @kidonng @stephan-hesselmann-by @dessalines. Guess what? I changed my mind! https://github.com/PatrickF1/fzf.fish/pull/120
